### PR TITLE
More projectile property streamlining and cleanups

### PR DESCRIPTION
**Pull request by reaperrr**
Another small step towards more consistency in property names, also some negligible performance tweaks and Lint rules to warn modders about invalid launch angles.

All yaml changes were performed with the respective upgrade rules.

### CODE REVIEW

**Comment by reaperrr** [comment]
> OpenRA.Utility(1,1): Error: OpenRA.Mods.Common.Lint.CheckAngle failed with exception: System.NullReferenceException: Object reference not set to an instance of an object

Does anyone have an idea what the hell is going on here?
Because this test *passes* on my local branch (I didn't make any changes since I pushed).

File: OpenRA.Mods.Common/Projectiles/AreaBeam.cs [hunk 1]
```diff
@@ -28,7 +28,7 @@ public class AreaBeamInfo : IProjectileInfo
 		public readonly WDist[] Speed = { new WDist(128) };
 
 		[Desc("The maximum duration (in ticks) of each beam burst.")]
-		public readonly int Duration = 10;
+		public readonly int BeamDuration = 10;
```
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/AreaBeam.cs]: We already know that this is a beam, so adding it again here doesn't help.

File: OpenRA.Mods.Common/Projectiles/LaserZap.cs [hunk 2]
```diff
@@ -23,6 +23,9 @@ namespace OpenRA.Mods.Common.Projectiles
 	[Desc("Not a sprite, but an engine effect.")]
 	public class LaserZapInfo : IProjectileInfo
 	{
+		[Desc("Should the beam be visually rendered? False = Beam is invisible.")]
+		public readonly bool RenderBeam = true;
```
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/LaserZap.cs]: A laser that isn't rendered isn't a laser at all (just a normal projectile), so isn't suited for this projectile type at all.

File: OpenRA.Mods.Common/Projectiles/LaserZap.cs [hunk 3]
```diff
@@ -100,17 +103,18 @@ public void Tick(World world)
...
 		{
-			if (wr.World.FogObscures(target) &&
-				wr.World.FogObscures(args.Source))
+			if (!info.RenderBeam && hitanim == null)
 				yield break;
 
-			if (ticks < info.BeamDuration)
+			var targetObscured = wr.World.FogObscures(target);
+			var sourceObscured = wr.World.FogObscures(args.Source);
+			if (!(targetObscured && sourceObscured) && ticks < info.BeamDuration)
```
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/LaserZap.cs]: This will cause the whole beam to be invisible if you get shot by an obelisk hidden under the fog, which looks much worse than what we have right now (I had already tried this a long time ago).

File: OpenRA.Mods.Common/Projectiles/AreaBeam.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Projectiles/AreaBeam.cs]: I intend to add `DamageDuration` in my next PR and felt this would be better to separate them, but if you prefer to rename `BeamDuration` in `LaserZap` to just `Duration` instead, I guess I can do that.

**Comment by reaperrr** [comment]
Updated.

However, I still need help with the lint check. For some reason it only works with positive values, negative angles always trigger the lint error (which is why I've disabled it for missiles for now).

**Comment by pchote** [comment]
`WAngle` is only defined for numbers between 0 and 1024.  The `new WAngle(-64)` in Missile.cs is internally converted to 960, which then fails your > 255 check.

**Comment by reaperrr** [comment]
> You check should be something like `return value > 255 && value < 768`;

Done, though I went for `value < 76*9*` because 768 should be the highest invalid value (1024 - 256 = 768).

File: OpenRA.Mods.Common/Projectiles/Missile.cs [hunk 4]
```diff
@@ -37,20 +37,12 @@ public class MissileInfo : IProjectileInfo
...
-		public readonly WDist MinimumLaunchSpeed = new WDist(-1);
-
-		[Desc("Maximum launch speed in WDist / tick. Defaults to Speed if -1.")]
-		public readonly WDist MaximumLaunchSpeed = new WDist(-1);
-
-		[Desc("Maximum projectile speed in WDist / tick")]
-		public readonly WDist Speed = new WDist(384);
+		[Desc("Projectile speed in WDist / tick, two values indicate launch speed and maximum speed,",
+			"three values represent minimum launch speed, maximum launch speed and maximum speed (in that order).")]
+		public readonly WDist[] Speed = { new WDist(384) };
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Projectiles/Missile.cs]: Notcertain if I'm a fan of this, this forces unneeded duplication at inherited ones.

**Comment by reaperrr** [comment]
Removed the first two laser-related commits, which are included in #12068 now.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by reaperrr** [comment]
Before I rebase this, any other opinions on the angle and speed field merging?

I see @GraionDilach's point, so I could live with dropping the commits in question.

**Comment by pchote** [comment]
I agree with @GraionDilach, so lets drop those commits.

**Comment by reaperrr** [comment]
Updated. 
Removed the commits in question and extended lint check to also emit errors when minimum launch angle exceeds maximum launch angle.

**Comment by reaperrr** [comment]
Updated.

Removed some syncing from `Bullet`, as it seemed excessive compared to `Missile`.

**Review by pchote — COMMENTED** [review]
Looks sensible overall, but I haven't looked in full detail or tested, so :+0.5: for now.

File: OpenRA.Mods.Common/Lint/CheckAngle.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+namespace OpenRA.Mods.Common.Lint
+{
+	class CheckAngle : ILintRulesPass
+	{
+		public void Run(Action<string> emitError, Action<string> emitWarning, Ruleset rules)
+		{
+			foreach (var weaponInfo in rules.Weapons)
+			{
+				var missile = weaponInfo.Value.Projectile as MissileInfo;
+				if (missile != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckAngle.cs]: Can you please pull this out into a `void CheckLaunchAngles(WAngle min, WAngle max)` to avoid duplication between these cases?

**Review by pchote — DISMISSED** [review]
Looks good after this fix

**Comment by reaperrr** [comment]
Updated the lint check.

Also made it slightly smarter, `maxAngle` is now only tested if it's not identical to `minAngle` to avoid duplicate errors when that single/identical angle is invalid.

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7f462506ec75487de697078778867d61862f5891
